src/webgpu/api/operation/queue/writeBuffer.spec.ts

```diff
@@ -1,11 +1,228 @@
-export const description = `
-TODO:
-- source.origin is unaligned
-- data given as ArrayBuffer
-- data given as each TypedArray variant
-`;
+export const description = 'Operation tests for GPUQueue.writeBuffer()';
 
+import { params, pbool, poptions } from '../../../../common/framework/params_builder.js';
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { range } from '../../../../common/framework/util/util.js';
 import { GPUTest } from '../../../gpu_test.js';
 
-export const g = makeTestGroup(GPUTest);
+const kTypedArrays = [
+  'Uint8Array',
+  'Uint16Array',
+  'Uint32Array',
+  'Int8Array',
+  'Int16Array',
+  'Int32Array',
+  'Float32Array',
+  'Float64Array',
+] as const;
+
+type WriteBufferSignature = {
+  bufferOffset: number;
+  data: readonly number[];
+  arrayType: typeof kTypedArrays[number];
+  useArrayBuffer: boolean;
+  dataOffset?: number;
+  dataSize?: number;
+};
+
+class F extends GPUTest {
+  calculateRequiredBufferSize(writes: WriteBufferSignature[]): number {
+    let bufferSize = 0;
+    // Calculate size of final buffer
+    for (const { bufferOffset, data, arrayType, useArrayBuffer, dataOffset, dataSize } of writes) {
+      const typeBuilder = self[arrayType];
+      let requiredSize = data.length * typeBuilder.BYTES_PER_ELEMENT;
+
+      // When passing data as an ArrayBuffer, dataOffset, dataSize use byte instead of number of
+      // elements
+      const bytesPerElement = useArrayBuffer ? 1 : typeBuilder.BYTES_PER_ELEMENT;
+      if (dataOffset) {
+        requiredSize -= dataOffset * bytesPerElement;
+      }
+      if (dataSize) {
+        requiredSize = Math.min(requiredSize, dataSize * bytesPerElement);
```

Comment:
It's a bit weird that we modify requiredSize to just set it here, maybe we could do:

```
bytesPerElement = ...
requiredSize = bufferOffset;
if (dataSize) {
  requiredSize += ...;
} else {
    requiredSize += data.length * BYTES_PER_ELEMENT  - (dataOffset || 0) * bytesPerElement
}
```